### PR TITLE
Add skip_tokenizer_init args.

### PR DESCRIPTION
**Pull request by gryffindor-rr**
Thank you for your contribution, we really appreciate it. The following instructions will help improve your pull request and make it easier to receive feedback. If there are any items you don't understand, don't worry. Just submit the pull request and ask the maintainers for help.

## Motivation

https://github.com/sgl-project/sglang/issues/924

## Modification

Added a skip_tokenizer_init to skip tokenizer and use the input_ids directly in tokenizer manager.

## Checklist

1. Ensure pre-commit `pre-commit run --all-files` or other linting tools are used to fix potential lint issues.
2. Confirm that modifications are covered by complete unit tests. If not, please add more unit tests for correctness.
3. Modify documentation as needed, such as docstrings or example tutorials.

### CODE REVIEW

**Review by zhyncs — COMMENTED** [review]
Hi @gryffindor-rr May you add a unit test for this? Thanks.

**Comment by zhyncs** [comment]
Hi @gryffindor-rr I've update the deps for SGLang, may you merge the latest main? Thanks.

File: python/sglang/srt/managers/detokenizer_manager.py [hunk 1]
```diff
@@ -69,70 +72,74 @@ async def handle_loop(self):
...
-                    # Streaming chunk: update the decode status
-                    if len(new_text) > 0 and not new_text.endswith("�"):
-                        s.decoded_text = s.decoded_text + new_text
-                        s.surr_offset = s.read_offset
-                        s.read_offset = len(s.decode_ids)
-                        new_text = ""
+            if self.tokenizer is None:
+                # Send BatchTokenIDOut if no tokenizer init'ed.
+                self.send_to_tokenizer.send_pyobj(recv_obj)
+            else:
```
Comment (Ying1123) [inline python/sglang/srt/managers/detokenizer_manager.py]: Do not move old code into a new nest level.
Can you try
```
            if self.tokenizer is None:
                # Send BatchTokenIDOut if no tokenizer init'ed.
                self.send_to_tokenizer.send_pyobj(recv_obj)
                continue
```

File: test/lang/test_skip_tokenizer_srt.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+from sglang.test.test_utils import DEFAULT_MODEL_NAME_FOR_TEST
+
+
+class TestSRTBackend(unittest.TestCase):
+    backend = None
+
+    @classmethod
+    def setUpClass(cls):
+        cls.backend = sgl.Runtime(model_path=DEFAULT_MODEL_NAME_FOR_TEST,
+                                  skip_tokenizer_init=True)
```
Comment (Ying1123) [inline test/lang/test_skip_tokenizer_srt.py]: Can you pass this test? I guess you cannot pass the tests.

File: test/lang/test_skip_tokenizer_srt.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import json
```
Comment (Ying1123) [inline test/lang/test_skip_tokenizer_srt.py]: For unittest, please do not use `test/lang`.
Can you add a unit test similar to this one https://github.com/sgl-project/sglang/blob/main/test/srt/test_srt_endpoint.py?

**Review by Ying1123 — CHANGES_REQUESTED** [review]
Can you run your code successfully?

I think you need also to change this part https://github.com/sgl-project/sglang/blob/db6e26d6efeff2ed2db1cb945a19a8d68a456ff3/python/sglang/srt/managers/tokenizer_manager.py#L484-L486.



Please fix the lint as well.

**Comment by gryffindor-rr** [comment]
> Hi @gryffindor-rr I've update the deps for SGLang, may you merge the latest main? Thanks.

done.

File: test/lang/test_skip_tokenizer_srt.py (hunk 3 above)
Comment (gryffindor-rr) [reply test/lang/test_skip_tokenizer_srt.py]: > For unittest, please do not use `test/lang`. Can you add a unit test similar to this one https://github.com/sgl-project/sglang/blob/main/test/srt/test_srt_endpoint.py?

done. moved to test/srt.

File: python/sglang/srt/managers/detokenizer_manager.py (hunk 1 above)
Comment (gryffindor-rr) [reply python/sglang/srt/managers/detokenizer_manager.py]: done.

File: test/lang/test_skip_tokenizer_srt.py (hunk 2 above)
Comment (gryffindor-rr) [reply test/lang/test_skip_tokenizer_srt.py]: modified the test and passed by set model_path to my local dir.

File: python/sglang/srt/server.py [hunk 4]
```diff
@@ -392,28 +392,29 @@ def _wait_and_warmup(server_args, pipe_finish_writer):
...
-    try:
-        for _ in range(server_args.dp_size):
-            res = requests.post(
-                url + "/generate",
-                json={
-                    "text": "The capital city of France is",
-                    "sampling_params": {
-                        "temperature": 0,
-                        "max_new_tokens": 8,
+    if not server_args.skip_tokenizer_init:
```
Comment (merrymercy) [inline python/sglang/srt/server.py]: let us keep this warmup test, but send token ids instead. We can send token ids `[10, 11, 12]`

File: python/sglang/srt/managers/tp_worker.py [hunk 5]
```diff
@@ -100,18 +100,24 @@ def __init__(
         )
 
         if is_multimodal_model(server_args.model_path):
```
Comment (merrymercy) [inline python/sglang/srt/managers/tp_worker.py]: check `server_args.skip_tokenizer_init` first. then check `if is_multimodal_model(server_args.model_path):`

File: python/sglang/srt/managers/tokenizer_manager.py [hunk 6]
```diff
@@ -92,24 +92,31 @@ def __init__(
             self.context_len = get_context_length(self.hf_config)
 
         if is_multimodal_model(self.model_path):
```
Comment (merrymercy) [inline python/sglang/srt/managers/tokenizer_manager.py]: check `if server_args.skip_tokenizer_init:` first.
```
if server_args.skip_tokenizer_init:
    self. tokenizer = self. processor = None
else:
    if is_multimodal_model(self.model_path):
        ...
    else:
        ...
```

**Review by merrymercy — CHANGES_REQUESTED** [review]
could you fix the lint errors? https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md#format-your-code

File: python/sglang/srt/managers/tp_worker.py (hunk 5 above)
Comment (gryffindor-rr) [reply python/sglang/srt/managers/tp_worker.py]: done.

File: python/sglang/srt/managers/tokenizer_manager.py (hunk 6 above)
Comment (gryffindor-rr) [reply python/sglang/srt/managers/tokenizer_manager.py]: done.

**Comment by zhyncs** [comment]
@gryffindor-rr Please resolve the CI issue.

**Comment by gryffindor-rr** [comment]
> could you fix the lint errors? https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md#format-your-code

done.

**Comment by gryffindor-rr** [comment]
> @gryffindor-rr Please resolve the CI issue.

done.

**Review by merrymercy — APPROVED** [review]
Great work!

**Review by Ying1123 — APPROVED** [review]

**Comment by merrymercy** [comment]
The tokens returned by this implementation is wrong. It is fixed by #1798

**Pull request merged**